### DISCUSSION TITLE
urql + React Native offline and persistence questions

### DISCUSSION THREAD

**Discussion by bombillazo** [urql-graphql/urql#1811] [Q&A] [answered]
Hello! First, I want to say THANK YOU to the urql team for the awesome work done with this library. I've read the docs, various discussion questions and issues, and repo code and I'm impressed with eh quality of the code and discussions. I would love for you all to have a Slack or Discord dev channel tbh. 😅

I am building a React Native app that I want to incorporate persisted caching and offline mode. My goals are the following:
- When querying, useQuery hook uses cache and fetches on background (I'm using `cache-and-network` fetchPolicy)
- All queried data is cached and persisted on the local device (using a custom secure storage)
- If the user closes the app, opening it again fetches data from the local device first
- If user is offline, urql uses `cache-only`
- If user comes back online, requests change `cache-and-network`
- Cache is not removed while offline mode
- Clear cache when user logs out

So far I have incorporated the `urql` library, incorporated the `offlineExchange` exchange from `@urql/exchange-graphcache`, implemented the storage interface with a custom storage. I see data gets copied to this store already. If this sounds unorthodox or convoluted and there is a better approach, I'm all ears! So here are my questions so far, I would greatly appreciate your inputs:

1. What is the difference between persistence and offline in the urql docs? Seems like they are 2 separate concepts.
2. Right now, when my component re-renders, the useQuery hook does not fetch data from the API after the first fetch. If data gets mutated outside of urql, urql still only uses the cache, even with `network-only`. I thought useQuery would run the fetch every time and use the fetchPolicy to manage the request. Am I misunderstanding how caching and fetching works?
3. How do I invalidate or clear all the cache? This might be necessary when login out users. By this I mean server data will no longer exist in cache nor in persisted storage.
4. Does `offlineExchange` automatically manage changing queryClient requestPolicy modes when going offline/online?

I apologize for the extensiveness, but clearing this up will allow me to move faster in development. Thanks for the help! 😄

**Comment by JoviDeCroock** [comment, marked as answer]
Persistence means we store the data across sessions, this can exist without offlineMode, offlineMode can't exist without persistence though. When we persist data from the `cacheExchange` subsequent visits to a site/app/... Will carry over the data from the last session. When we add offline on top of this then we'll also carry over the mutations/... the user has done while being offline, this way we can dispatch all of these operations again and achieve the previous `optimistic` state.

When using a `cache-first` policy we'll always hit the cache and prefer not to send a request if the cache gives us a _complete_ answer, if it's partial or incomplete we'll dispatch a request. With `network-only` there should always be a request unless your `offlineExchange` tells us you're offline, then it will probably be a cached result 😅 (not too sure on that one currently).

There are two ways of cache clearing, the persistence layer can be cleared by calling `.clear()` on the storage object. Our in-memory cache is commonly cleared by recreating the urql-client.

In the `offlineExchange` we keep a queue that tracks the requests that failed with an "offline-error" https://github.com/FormidableLabs/urql/blob/main/exchanges/graphcache/src/offlineExchange.ts#L60 when the user goes back online we'll attempt to flush this failed queue and dispatch the operations the user had before.

**Reply by bombillazo** [reply]
Thanks for the clarification on points 1, 3 and 4.

On point 2, I am still struggling to understand some things. I read this in the docs:
`The useQuery hook updates and executes queries whenever its inputs, like the query or variables, change but in some cases, we may find that we need to programmatically trigger a new query.`

I am using a functional component in React. Does this mean that once I mount my component, and use `useQuery` inside my function, if I don't change the inputs and don't unmount the component, it will _never_ refetch data? Even if I re-render my component, another fetch never happens until the component unmounts and mounts again? What if another component uses the same query+inputs? Will it fetch data or see the cached data and not try to run the query?

This is the behavior I seem to be getting with my app, and I want to confirm it is the expected behavior.

I might get why one would not want to fetch data on every rerender, but this is important to me because if I can use the fetchPolicy exchange to switch from `cache-first`  to `cache-and-network`, I can limit said network usage. But if I only fetch mount and on input changes, that messes up for components that are persistent across screens or views. I would have to programmatically refresh the data.

**Reply by JoviDeCroock** [reply]
This is expected with cache-first yes, you might want to look into the requestPolicyExchange, this automatically upgrades cache-first queries with a certain age. Also when you trigger a change to the argument or use the second argument returned from usequery you can force a refetch.

https://github.com/FormidableLabs/urql/tree/main/exchanges/request-policy

**Reply by bombillazo** [reply]
And on point 3, and I am doing some tests and I am having trouble with these 2:

- [x] **_Cache is not removed while offline mode_**
- [ ] _Clear cache when user logs out_

`Cache is not removed while offline mode` works _so_ well that while my app is offline, I clear both the in-memory cache (recreating the client) and persisted cache (clearing the cache object store directly via its own `clearStore` API call, I saw no clear method specified in the `StorageAdapter` interface) and unmount the only component using the data and I still get data from someplace where I would expect no values. I inspecting the storage (still offline) and somehow the data is there. I'm thinking there is some internal data object being "cached" inside graphcache that is rehydrating the storage or in-memory cache. 

This leads me to not being able to reliably clear the cache and guarantee previous data is leaked to the nest usage

I tried this with both `cacheExchange` and `offlineExchange`  from the graphcache package (confusing to have 2 exchanges with the same name btw, e.i the og default `cacheExchange` and the graphcache `cacheExchange`). Any clue what may be preventing a total cache clear?

**Reply by JoviDeCroock** [reply]
I would need to see a reproduction but normally when the storage-object is cleared _and_ you recreate the client after it should all be good. An example recreation could look like:

```jsx
function GraphQLProvider(props) {
  const [client, setClient] = useState(/* a function that creates your client */);
  const resetClient = /* function that resets the client*/
  return Provider client={client} {children}
}
```

**Reply by bombillazo** [reply]
I followed that approach from another answer, quite clever!

```js
export const URQLProvider = ({ children }) => {
  const [client, setClient] = useState(makeClient());
  return (
    <URQLContext.Provider
      value={{
        resetClient: () => setClient(makeClient()),
      }}
    >
      <Provider value={client}>{children}</Provider>
    </URQLContext.Provider>
  );
};

...

// resetting local store and urql client
  URQLStore.clearStore(); // store object used by 
  resetClient();
```

Still seems to not be clearing for some reason. I do notice in the [custom storage](https://formidable.com/open-source/urql/docs/graphcache/offline/#custom-storages) docs, there is this internal variable in the storage interface:

```js
const makeLocalStorage = () => {
  const cache = {}; // cache variable?!

  return {
    writeData(delta) {
      return Promise.resolve().then(() => {
        Object.assign(cache, delta);
        localStorage.setItem('data', JSON.stringify(cache));
      });
    },
```
Could that be the culprit? I'm pretty much using the same code but with another storage library. How would one clear that value when resetting the client?

**Reply by JoviDeCroock** [reply]
Are you not resetting the `cache` and `storage`? Like if you use an already created v ariable then you are reusing instances which in that case it's expected that your cache is being reused, you should recreate the exchange and storage.

Make sure `makeClient` also resets the storage and cache.
